harness/determined/common/api/certs.py

```diff
@@ -0,0 +1,220 @@
+            # (see https://docs.python.org/3/library/tempfile.html#tempfile.NamedTemporaryFile)
+            fd, combined_path = tempfile.mkstemp(prefix="det-master-cert-")
+            atexit.register(os.unlink, combined_path)
+
+            with open(fd, "wb") as out:
+                with open(certifi.where(), "rb") as base_certs:
+                    out.write(base_certs.read())
+                out.write(b"\n")
+                out.write(cert_pem.encode("utf8"))
+
+            self._bundle = combined_path
+
+    @property
+    def bundle(self) -> Union[None, str, bool]:
+        """
+        The path to a file containing an SSL certificate to trust specifically for the master, if
+        any, or False to disable cert verification entirely. If set to a path, it should always be a
+        temporary file that we own and can delete.
+        """
+        return self._bundle
+
+    @property
+    def name(self) -> Optional[str]:
+        """
+        The name we use to verify the master certificate.
+        """
+        return self._name
+
+
+cli_cert = None  # type: Optional[Cert]
+
+
+class _BrokenCertStore(Exception):
+    pass
+
+
+def _load_cert_store(path: pathlib.Path) -> Dict[str, str]:
+    if not path.exists():
+        return {}
+    try:
+
+        with path.open() as f:
+            content = f.read()
+
+        try:
+            store = json.loads(content)
+        except json.JSONDecodeError:
+            raise _BrokenCertStore()
+
+            if not isinstance(store, dict):
+                raise _BrokenCertStore()
+
+        # Store must be a dictionary.
+        if not isinstance(store, dict):
+            raise _BrokenCertStore()
+
+        # All keys are url's, all values are pem-encoded certs.
+        for k, v in store.items():
+            if not isinstance(k, str):
+                raise _BrokenCertStore()
+            if not isinstance(v, str):
+                raise _BrokenCertStore()
+
+    except _BrokenCertStore:
+        path.unlink()
+        return {}
+
+    return cast(Dict[str, str], store)
+
+
+@contextlib.contextmanager
+def _modifiable_store(path: pathlib.Path) -> Iterator[Dict["str", "str"]]:
+    """
+    Yields the appropriate store that can be modified, and the modified result will be written
+    back to file.
+
+    The modified store is also saved to the local object.
+    """
+    path = path
+    path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
+    # Decide on paths for a lock file and a temp files (during writing).
+    temp = pathlib.Path(str(path) + ".temp")
+    lock = pathlib.Path(str(path) + ".lock")
+
+    with filelock.FileLock(lock):
+        store = _load_cert_store(path)
+
+        # No need for try/finally, because we don't update the file after failures.
+        yield store
+
+        with temp.open("w") as f:
+            json.dump(store, f, indent=4, sort_keys=True)
+        temp.rename(path)
+
+
+def get_cert(path: pathlib.Path, url: str) -> Optional[str]:
```

Comment:
Does `get_cert` just return the path to the cert file?

Reply:
No, it is returning the content of the certificate.  Added a docstring.